File: engine/include/xjmusic/entities/meme/ParseUnique.java
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.

package io.xj.model.meme;

import java.util.std::vector;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 Meme Matcher for Unique-Memes
 <p>
 Parse any meme to test if it's valid, and extract its features
 <p>
 Artist can add `$MEME` so only one is chosen https://github.com/xjmusic/workstation/issues/219
 */
class ParseUnique {
  static final Pattern rgx = Pattern.compile("^\\$(.+)$");
  final std::string body;
  final boolean isValid;

  ParseUnique(std::string raw) {
    Matcher matcher = rgx.matcher(raw);

    if (!matcher.find()) {
      body = null;
      isValid = false;
      return;
    }

    body = matcher.group(1);
    if (java.util.Objects.isNull(body) || body.length() == 0) {
      isValid = false;
      return;
    }

    isValid = true;
  }

  public static ParseUnique fromString(std::string raw) {
    return new ParseUnique(raw);
  }

  public boolean isViolatedBy(ParseUnique target) {
    return isValid && target.isValid && Objects.equals(body, target.body);
  }

  public boolean isAllowed(std::vector<ParseUnique> memes) {
    for (var meme : memes)
      if (isViolatedBy(meme)) return false;
    return true;
  }
}

File: engine/src/meme/ParseUnique.cpp
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.

#include "xjmusic/meme/ParseUnique.h"

using namespace XJ;

ParseUnique::ParseUnique(const std::string &raw) {
  std::smatch matcher;
  valid = std::regex_search(raw, matcher, rgx);

  if (!valid) {
    body = "";
    valid = false;
    return;
  }

  body = matcher[1].str();
  if (body.empty()) {
    valid = false;
    return;
  }

  valid = true;
}

ParseUnique ParseUnique::fromString(const std::string &raw) {
  return ParseUnique(raw);
}

bool ParseUnique::isViolatedBy(const ParseUnique &target) const {
  return valid && target.valid && body == target.body;
}

bool ParseUnique::isAllowed(const std::vector<ParseUnique> &memes) {
  return std::all_of(memes.begin(), memes.end(), [this](const ParseUnique &meme) {
    return !isViolatedBy(meme);
  });
}

const std::regex ParseUnique::rgx("^\\$(.+)$");
